[Preview] Functions (#1479)
momentohq/public-dev-docs: docs/functions/index.mdx

---
sidebar_position: 1
sidebar_label: Momento Functions
title: Momento Functions
description: Process real-time data at massive scale with high-performance Momento Functions.
hide_title: true
slug: /functions
keywords:
  - momento
  - functions
  - wasm
  - real-time processing
  - event-driven
  - serverless
---

# Momento Functions

> *Momento Functions* are a high-speed compute layer designed for processing massive amounts of real-time data. Whether you're aggregating IoT sensor data, handling financial transactions, or enriching event streams with AI, *Momento Functions* allow you to process, transform, and act on data instantly—without managing infrastructure.

Reviewer: ```suggestion
*Momento Functions* is a lightweight compute layer designed for processing real-time data. Whether you're aggregating IoT sensor data, recording financial transactions, or enriching event streams with AI, *Momento Functions* allows you to process, transform, and act on data instantly—without managing infrastructure.
```
Looks good, but I'd tighten this up a little further.

> Built for extreme scale and performance, functions offer low-latency execution and shared global memory across all running function instances. Unlike general-purpose serverless compute, they do not execute in isolation, allowing all instances of a function to share state for efficient data processing.

Reviewer: I think this wording isn't quite right for this implementation detail, and I'm struggling to come up with an approach that succinctly captures the differentiation's semantics. It might be best to either pull this paragraph or iterate on it more.

```mermaid
mindmap
  root((Momento Functions))
    (Fast Execution)

>       (5ms SLA for ultra-fast processing)

Reviewer: I don't think we're offering an SLA. Some functions will take much longer to execute than 5ms, and it's largely outside of Momento's control how long they take. We can't make Bedrock go any faster than it does, and we can't exceed the universe's rate of information propagation (C).
Maybe we can say something about sub-millisecond cold starts, if you want to add a latency number? I think people typically associate that number with the right technical detail.

>       (10x faster, 10x cheaper)

Reviewer: Than...? 😄

    (Shared Global Memory)

>       (All function instances share state)

Reviewer: This is sort of the case, and I'm not sure how far we'll go down the path of pretending that an in-memory cache is global. We could make it global and transactional, but that would slow down functions tremendously!

      (Persistent between executions)
    (Batch Processing)
      (Dynamically process batched events)
      (Seamless data aggregation)
    (Seamless Integrations)
      (Amazon S3)
      (Momento Topics)
      (Momento Cache)

>       (Amazon Bedrock)

Reviewer: Maybe you want to either alphabetize or group the integrations by top-level owner?

```

## What makes Momento Functions different?
Traditional serverless platforms execute functions in isolation, requiring external coordination for batching, state sharing, and data transformations. Momento Functions remove these limitations:

* Global memory shared across all function instances enables stateful execution without an external database.
* Automatic batching allows functions to process multiple related events in a single execution cycle.

> * No concurrency limits means functions can instantly scale to handle millions of invocations per second without cold starts.

Reviewer: I'm not sure we'll have no concurrency limits. I think it's going to be easy to set high limits, but we can't sustain, e.g., a trillion concurrent requests on each frontend host.
